SCons: `silence_msvc` implementation issues

### Tested versions

There are issues with the `silence_msvc` implementation as a result of changes made in PRs #90626 and #90551.


### System information

Windows 11, Godot master, SCons 4.7.0

### Issue description

There are issues with the `silence_msvc` implementation as a result of changes made in PRs #90626 and #90551.

A similar PR was recently opened on the SCons project that linked to the one of the merged PRs above.  A detailed explanation that goes beyond the content presented below has not been posted on SCons yet.

Environment: Windows 11, Godot master (2024/05/10), SCons 4.7.0.

#### Executive Summary

As currently implemented:
 1. *The first line of an msvc linker warning, error, and diagnostic message may be erroneously deleted.*
    
    This is due to a combination of the inherent assumptions made in the Godot implementation, the differences in output stream behavior between the msvc compiler and msvc linker, and the differences in timing when the filtered records are produced.
    
 2. *The binary read of the temporary stdout file followed by a decoded write to the stderr stream will likely contain extra newline characters for each line.*
 
    The binary read of the temporary stdout file contains *CRLF* (i.e., `\r\n`) line endings.  The *CRLF* sequence remains after decoding.  This effectively causes an extra new line for each line of content written.
 
 3. *There may be unnecessary unicode replacement characters introduced in the written stderr content due to the implemented encoding.*
 
    The encoding selected is that of python's stdout stream.  However, the temporary stdout file is populated by the output of the msvc compiler and linker commands.  SCons uses encoding "oem" in some cases when processing msvc command output.

#### Suggested Changes

At a bare minimum, the following changes are suggested for the current Godot implementation:
1. Delete the first record *only* for the msvc compiler.  The risks with erroneously deleting the wrong msvc linker record far outweigh the benefit of suppressing a couple lines of output.
2. Replace the *CRLF* sequence with a *LF* in the decoded stream.
3. Change the encoding from `sys.stdout.encoding` to  `"oem"`.

Suggested changes to `godot/platform/windows/detect.py` master:
```diff
    if env["silence_msvc"]:
        from tempfile import mkstemp

        old_spawn = env["SPAWN"]

        def spawn_capture(sh, escape, cmd, args, env):
            # We only care about cl/link, process everything else as normal.
            if args[0] not in ["cl", "link"]:
                return old_spawn(sh, escape, cmd, args, env)

            tmp_stdout, tmp_stdout_name = mkstemp()
            os.close(tmp_stdout)
            args.append(f">{tmp_stdout_name}")
            ret = old_spawn(sh, escape, cmd, args, env)

            try:
                with open(tmp_stdout_name, "rb") as tmp_stdout:
                    # First line is always bloat, subsequent lines are always errors. If content
                    # exists after discarding the first line, safely decode & send to stderr.
-                   tmp_stdout.readline()
+                   if args[0] in ["cl"]:
+                       tmp_stdout.readline()
                    content = tmp_stdout.read()
                    if content:
-                       sys.stderr.write(content.decode(sys.stdout.encoding, "replace"))
+                       sys.stderr.write(content.decode("oem", "replace").replace("\r\n", "\n"))
                os.remove(tmp_stdout_name)
            except OSError:
                pass

            return ret

        env["SPAWN"] = spawn_capture
```

Additional concerns such as potentially causing a user stream redirection argument to be ignored are beyond the scope of this post.

#### MSVC Compiler and Linker Default Output Streams

| Default Output Stream | Category | cl.exe | link.exe |
| :---------- | :----: | :----: | :------: |
| Executable banner messages | Diagnostic | $\color{red}\textit{stderr}$ | $\color{red}\textit{stdout}$ |
| Command-line warning messages | Warning | $\color{red}\textit{stderr}$ | $\color{red}\textit{stdout}$ |
| Command-line error messages | Error | $\color{red}\textit{stderr}$<sup>1</sup> | $\color{red}\textit{stdout}$ |
| Diagnostic messages | Diagnostic | stdout | stdout |
| Warning messages | Warning | stdout | stdout |
| Error messages | Error | stdout | stdout |
| User pragma messages | Diagnostic | stdout | $\color{gray}\textit{n/a}$ |
| User command wrapper<sup>2</sup> | Diagnostic | stderr and/or stdout | stderr and/or stdout |

Table Notes:
1. VS 2022 introduced compiler option `/options:strict` which treats unrecognized compiler options as errors.
2. Due to the current implementation of using the program name without an extension to invoke the msvc compiler (`cl`) and msvc linker (`link`), it is trivial to insert a user command wrapper (i.e., "man-in-the-middle") batch file for the compiler (`cl.bat`) and linker (`link.bat`) that calls the compiler executable (`cl.exe`) and linker executable (`link.exe`) when manually configuring the Godot MSVC environment (i.e., "VCVARS" method).  Such a command wrapper may write to stdout and stderr, *unwittingly* redirect the command stderr stream to stdout, and otherwise change the expected record sequence.

#### Implementation Assumptions

> First line is always bloat, subsequent lines are always errors. If content  
  exists after discarding the first line, safely decode & send to stderr.

Most of the immediate concerns have to do with deleting the first line of the msvc linker output.
  
MSVC compiler:

* The diagnostic message containing the file name appears to be written after the command-line options are evaluated and before the source files are processed.

* The compiler command line accepts a list of files.  Diagnostic messages containing the source file name are issued for *each* source file processed.  This is useful information outside of SCons when compiling multiple files with a single command.  The file name diagnostic message is not as useful with SCons as the compiler command is written to stdout and only a single source file is compiled per command.
  
* The first line of stdout content is not *always* the diagnostic message containing the file name due to command-line option warnings possibly being treated as errors in VS2022 and later (i.e., the file name diagnostic message may not be present).

* *If an external command wrapper were to redirect the compiler command stderr to stdout, then the first line of the banner message, command-line warning, or command-line error would be erroneously deleted.*

* Subsequent lines following the first line are not *always errors*.

  There are diagnostic options such as `/Bt`, `/showIncludes` and source code `pragma` messages which produce stdout messages that are neither errors or warnings.
  
* If the command output ever interleaves writes to stdout and stderr, the resulting output would not be in the correct order (i.e., command stderr written directly, followed by filtered stdout).  While largely theoretical, this can be demonstrated with a command wrapper.
  
MSVC linker:

* The linker does not *always* produce a diagnostic message.

  I believe that the "Creating library ..." message is produced as a result of the linker detecting exported symbols.
  
* The banner message, command-line warnings, command-line errors, diagnostic messages, warnings, and errors (e.g., multiply defined symbols) ***may be produced before*** the expected diagnostic message.  In these cases, deleting the first line is *erroneous*.

* For linker commands that ***do not produce*** the expected diagnostic message, deleting the first line is guaranteed to be *erroneous*.

* Subsequent lines following the first line are not *always errors*.

  There are diagnostic options such as `/time` and `/VERBOSE` which produce stdout messages that are neither errors or warnings.

The compiler and linker may produce other diagnostic messages like "Generating Code ..." and "Compiling ..." which could affect the expected record sequence.

Opinions will vary with regard to writing diagnostic messages (e.g., source code pragma messages) to stderr.

*In my experience, deleting an output record based solely on record count without verifying the record content is asking for trouble.  It does not help that the deleted content is not logged anywhere.*

#### Examples

The output fragments shown below are from a locally modified copy of the Godot source that writes the filtered line to stderr.

* Binary read *CRLF* sequences:

  * Extra new lines:

    `ccflags="/Bt" linkflags="/time"`

    ```
    cl ...  /Bt ...
    *** FILTER CL ***: console_wrapper_windows.cpp
    time(C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.39.33519\bin\HostX64\x64\c1xx.dll)=0.433s

    time(C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.39.33519\bin\HostX64\x64\c2.dll)=0.064s

    ```
  
    ```
    link ... /time ...
    *** FILTER LINK ***:    Creating library bin\godot.windows.template_release.x86_64.lib and object bin\godot.windows.template_release.x86_64.exp
    LibDef: Total time = 0.000s

      OptRef: Total time = 0.187s

      OptIcf: Total time = 2.453s

    Pass 1: Interval #1, time = 7.063s

    Pass 2: Interval #2, time = 2.812s

    Final: Total time = 9.875s

    ```
  
    *Erroneous deletion*:
    ```
    link ... /time ...
    *** FILTER LINK ***:   OptRef: Total time = 0.000s
      OptIcf: Total time = 0.000s

    Pass 1: Interval #1, time = 0.563s

    Pass 2: Interval #2, time = 0.000s

    Final: Total time = 0.563s

    ```

  * With *CRLF* replaced with *LF*:

    `ccflags="/Bt" linkflags="/time"`

    ```
    cl ... /Bt ...
    *** FILTER CL ***: console_wrapper_windows.cpp
    time(C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.39.33519\bin\HostX64\x64\c1xx.dll)=0.385s
    time(C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.39.33519\bin\HostX64\x64\c2.dll)=0.025s
    ```
  
    ```
    link ... /time ...
    *** FILTER LINK ***:    Creating library bin\godot.windows.template_release.x86_64.lib and object bin\godot.windows.template_release.x86_64.exp
    LibDef: Total time = 0.016s
      OptRef: Total time = 0.172s
      OptIcf: Total time = 2.531s
    Pass 1: Interval #1, time = 7.266s
    Pass 2: Interval #2, time = 3.359s
    Final: Total time = 10.625s
    ```
  
    *Erroneous deletion:*
    ```
    link ... /time ...
    *** FILTER LINK ***:   OptRef: Total time = 0.000s
      OptIcf: Total time = 0.000s
    Pass 1: Interval #1, time = 0.546s
    Pass 2: Interval #2, time = 0.000s
    Final: Total time = 0.546s
    ```
  
* Encoding (with *CRLF* replacement) shown for German:

  * Original encoding:
  
    `werror=yes linkflags="/SUBSYSTEM:CONSOLE,1.0"`
  
    *Erroneous deletion:*
    ```
    link ... /SUBSYSTEM:CONSOLE,1.0 ...
    *** FILTER LINK ***: LINK : warning LNK4010: Ung\ufffdltige Versionsnummer 1.0; Standardversion des Subsystems wird angenommen.
    LINK : error LNK1218: Warnung wird als Fehler interpretiert; es wurde keine Ausgabedatei generiert.
    ```
  
    Filtered line contains: `Ung\ufffdltige`.
  
  * Modified encoding:

    `werror=yes linkflags="/SUBSYSTEM:CONSOLE,1.0"`

    *Erroneous deletion:*
    ```
    link ... /SUBSYSTEM:CONSOLE,1.0 ...
    *** FILTER LINK ***: LINK : warning LNK4010: Ungültige Versionsnummer 1.0; Standardversion des Subsystems wird angenommen.
    LINK : error LNK1218: Warnung wird als Fehler interpretiert; es wurde keine Ausgabedatei generiert.
    ```

    Filtered line contains: `Ungültige`.
 
 
* Additional erroneous deletion examples:

  * `werror=yes linkflags="/CmdLineOpt"`
  
    ```
    link ... /CmdLineOpt ...
    *** FILTER LINK ***: LINK : warning LNK4044: unrecognized option '/CmdLineOpt'; ignored
    LINK : error LNK1218: warning treated as error; no output file generated
    ```
    
  * `werror=no linkflags="/CmdLineOpt"`
  
    ```
    link ... /CmdLineOpt ...
    *** FILTER LINK ***: LINK : warning LNK4044: unrecognized option '/CmdLineOpt'; ignored
       Creating library bin\godot.windows.template_release.x86_64.lib and object bin\godot.windows.template_release.x86_64.exp
    ```
    
    ```
    link ... /CmdLineOpt
    *** FILTER LINK ***: LINK : warning LNK4044: unrecognized option '/CmdLineOpt'; ignored
    Building compilation database compile_commands.json
    ```
    
  * `linkflags="/VERBOSE:LIB"`
 
    ```
    link ... /VERBOSE:LIB ...
    *** FILTER LINK ***: 
    Searching libraries
        Searching C:\Program Files (x86)\Windows Kits\10\Lib\10.0.22621.0\um\x64\winmm.lib:
        ...
        Searching C:\Program Files (x86)\Windows Kits\10\Lib\10.0.22621.0\um\x64\uuid.lib:
 
    Finished searching libraries
       Creating library bin\godot.windows.template_release.x86_64.lib and object bin\godot.windows.template_release.x86_64.exp
    ...   
    ```

#### Postscript

*I believe there is an issue with the current Godot implementation of msvc auto-detection (i.e., `setup_msvc_auto`).*

*By forcing SCons to re-initialize the environment and then reloading the msvc compiler tool, some Godot SConstruct command-line options for the msvc compiler that were configured earlier will be lost due to being overwritten by SCons when re-initialized.  At a minimum, this likely affects `ccflags`, `CCFLAGS`, and the system path.  There may be other consequences.*


### Steps to reproduce

See the Examples section above for command-line options that produce anomalous behavior.

The following should produce a linker warning that is suppressed/deleted by `silence_msvc` implementation.
```
scons platform=windows target=template_release tests=false module_text_server_fb_enabled=yes d3d12=no debug_symbols=no verbose=yes warnings=extra werror=yes linkflags="/CmdLineOpt"
```

Note: configuring the msvc environment manually (i.e., from the command-line and relying on Godot's VCVARS version) is necessary as the msvc auto-detection seems to ignore the ccflags and linkflags command-line options.  See the Postscript above for more details.

### Minimal reproduction project (MRP)

N/A

Edit:
* Fixed typographical error: VSVARS to VCVARS

--- a/platform/windows/detect.py
+++ b/platform/windows/detect.py
@@ -1,5 +1,6 @@
 import methods
 import os
+import re
 import subprocess
 import sys
 from methods import print_warning, print_error
@@ -304,6 +305,7 @@ def setup_msvc_manual(env: "SConsEnvironment"):
     print("Using VCVARS-determined MSVC, arch %s" % (env_arch))
 
 
+# FIXME: Likely overwrites command-line options for the msvc compiler. See #91883.
 def setup_msvc_auto(env: "SConsEnvironment"):
     """Set up MSVC using SCons's auto-detection logic"""
 
@@ -386,15 +388,28 @@ def configure_msvc(env: "SConsEnvironment", vcvars_msvc_config):
 
     env["MAXLINELENGTH"] = 8192  # Windows Vista and beyond, so always applicable.
 
-    if env["silence_msvc"]:
+    if env["silence_msvc"] and not env.GetOption("clean"):
         from tempfile import mkstemp
 
+        # Ensure we have a location to write captured output to, in case of false positives.
+        capture_path = methods.base_folder_path + "platform/windows/msvc_capture.log"
+        with open(capture_path, "wt"):
+            pass
+
         old_spawn = env["SPAWN"]
+        re_redirect_stream = re.compile(r"^[12]?>")
+        re_cl_capture = re.compile(r"^.+\.(c|cc|cpp|cxx|c[+]{2})$", re.IGNORECASE)
+        re_link_capture = re.compile(r'\s{3}\S.+\s(?:"[^"]+.lib"|\S+.lib)\s.+\s(?:"[^"]+.exp"|\S+.exp)')
 
         def spawn_capture(sh, escape, cmd, args, env):
             # We only care about cl/link, process everything else as normal.
             if args[0] not in ["cl", "link"]:
                 return old_spawn(sh, escape, cmd, args, env)
+
+            # Process as normal if the user is manually rerouting output.
+            for arg in args:
+                if re_redirect_stream.match(arg):
+                    return old_spawn(sh, escape, cmd, args, env)
 
             tmp_stdout, tmp_stdout_name = mkstemp()
             os.close(tmp_stdout)
@@ -402,16 +417,34 @@ def configure_msvc(env: "SConsEnvironment", vcvars_msvc_config):
             ret = old_spawn(sh, escape, cmd, args, env)
 
             try:
-                with open(tmp_stdout_name, "rb") as tmp_stdout:
-                    # First line is always bloat, subsequent lines are always errors. If content
-                    # exists after discarding the first line, safely decode & send to stderr.
-                    tmp_stdout.readline()
-                    content = tmp_stdout.read()
-                    if content:
-                        sys.stderr.write(content.decode(sys.stdout.encoding, "replace"))
+                with open(tmp_stdout_name, encoding="oem", errors="replace") as tmp_stdout:
+                    lines = tmp_stdout.read().splitlines()
                 os.remove(tmp_stdout_name)
             except OSError:
                 pass
+
+            # Early process no lines (OSError)
+            if not lines:
+                return ret
+
+            is_cl = args[0] == "cl"
+            content = ""
+            caught = False
+            for line in lines:
+                # These conditions are far from all-encompassing, but are specialized
+                # for what can be reasonably expected to show up in the repository.
+                if not caught and (is_cl and re_cl_capture.match(line)) or (not is_cl and re_link_capture.match(line)):
+                    caught = True
+                    try:
+                        with open(capture_path, "a") as log:
+                            log.write(line + "\n")
+                    except OSError:
+                        print_warning(f'Failed to log captured line: "{line}".')
+                    continue
+                content += line + "\n"
+            # Content remaining assumed to be an error/warning.
+            if content:
+                sys.stderr.write(content)
 
             return ret
 

